### PR TITLE
feat(datasets): Add Hugging Face datasets

### PR DESCRIPTION
**Pull request by astrojuanlu**
## Description

As per title 🤗 

cc @dannyrfar @kevin-koga-mckinsey @matescharnitzky @tomershor @pymousse @belavaishali @khangjrakpamarjun

## Development notes

This is still missing tests and docstrings. Ready for review code-wise.

## Checklist

- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the relevant `RELEASE.md` file
- [ ] Added tests to cover my changes

### CODE REVIEW

File: kedro-datasets/setup.py [hunk 1]
```diff
@@ -92,6 +98,7 @@ def _collect_requirements(requires):
     "databricks": _collect_requirements(databricks_require),
     "geopandas": _collect_requirements(geopandas_require),
     "holoviews": _collect_requirements(holoviews_require),
+    "huggingdace": _collect_requirements(huggingface_require),
```
Comment (noklam) [inline kedro-datasets/setup.py:101]:
```suggestion
    "huggingface": _collect_requirements(huggingface_require),
```

**Review by noklam — COMMENTED** [review]
notfamiliar with the hf pipeline, would there be some example of how to use this?

**Comment by astrojuanlu** [comment]
> notfamiliar with the hf pipeline, would there be some example of how to use this?

We do have an example we can share - where would be the best place to put it? A `README.md` in that directory, the module docstring, the classes docstrings?

File: kedro-datasets/kedro_datasets/huggingface/__init__.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+"""Provides interface to Unity Catalog Tables."""
```
Comment (astrojuanlu) [inline kedro-datasets/kedro_datasets/huggingface/__init__.py:1]: Woops, not really 😄

**Comment by noklam** [comment]
> We do have an example we can share - where would be the best place to put it? A README.md in that directory, the module docstring, the classes docstrings?

README.md in the dataset is rare and I believe it won't be index in the kedro-docs? Normally we provide both YAML and Python API example in the docstrings.

File: kedro-datasets/kedro_datasets/huggingface/hugging_face_dataset.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
+from __future__ import annotations
+
+from typing import Any
+
+from datasets import load_dataset
```
Comment (merelcht) [inline kedro-datasets/kedro_datasets/huggingface/hugging_face_dataset.py:5]: that's a confusing library name 😅 for a sec I thought this came from inside `kedro-datasets`..

**Review by merelcht — COMMENTED** [review]
The code looks fine, but it's hard to review without tests and examples since I don't know much about HuggingFace.

**Comment by noklam** [comment]
I've converted this back to draft since I see that the tests are still empty, feel free to re-request review when it's ready.

File: kedro-datasets/kedro_datasets/huggingface/hugging_face_dataset.py (hunk 3 above)
Comment (astrojuanlu) [reply kedro-datasets/kedro_datasets/huggingface/hugging_face_dataset.py:5]: 🙈

File: kedro-datasets/docs/source/conf.py [hunk 4]
```diff
@@ -90,6 +91,9 @@
     "kedro_docs_style_guide.md",
 ]
 
+intersphinx_mapping = {
```
Comment (astrojuanlu) [inline kedro-datasets/docs/source/conf.py:94]: @stichbury adding this here too, see below

File: kedro-datasets/kedro_datasets/huggingface/hugging_face_dataset.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+from huggingface_hub import HfApi
+from kedro.io import AbstractVersionedDataset as AbstractVersionedDataSet
+from kedro.io import DatasetError as DataSetError
+
+
+class HFDataset(AbstractVersionedDataSet):
+    """``HFDataset`` loads Hugging Face datasets
+    using the `datasets <https://pypi.org/project/datasets>`_ library.
+
+    Example usage for the :doc:`YAML API <kedro:data/data_catalog_yaml_examples>`:
```
Comment (astrojuanlu) [inline kedro-datasets/kedro_datasets/huggingface/hugging_face_dataset.py:15]: @stichbury ☝🏽

**Comment by astrojuanlu** [comment]
Rendered result:

- https://kedro--344.org.readthedocs.build/projects/kedro-datasets/en/344/kedro_datasets.huggingface.HFTransformerPipelineDataset.html
- https://kedro--344.org.readthedocs.build/projects/kedro-datasets/en/344/kedro_datasets.huggingface.HFDataset.html

**Review by stichbury — APPROVED** [review]
Approved from docs perspective. Exciting 👯

File: kedro-datasets/kedro_datasets/huggingface/transformer_pipeline_dataset.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
+from __future__ import annotations
+
+import typing as t
+from typing import Any
```
Comment (noklam) [inline kedro-datasets/kedro_datasets/huggingface/transformer_pipeline_dataset.py:4]: Can we just keep one here?

**Review by noklam — APPROVED** [review]
Generally looks good!

File: kedro-datasets/kedro_datasets/huggingface/transformer_pipeline_dataset.py (hunk 6 above)
Comment (astrojuanlu) [reply kedro-datasets/kedro_datasets/huggingface/transformer_pipeline_dataset.py:4]: Ugh, sure

**Pull request merged**